File: 2022/tom/day-1/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
)


func main() {
	data := readFile("input.txt")
	step1Calories := step1(data)
	fmt.Println(step1Calories)
	step2Calories := step2(data)
	fmt.Println(step2Calories)
}

func readFile(fileName string) []string {
	data, _ := os.Open(fileName)
	scanner := bufio.NewScanner(data)

	var dataLines []string
	for scanner.Scan() {
		line := scanner.Text()
		dataLines = append(dataLines, line)
	}
	return dataLines
}

func step1(data []string) int {
	var topCalories int = 0
	var currentCalories int = 0
	for _,line := range data {
		if (line == "") {
			if topCalories < currentCalories {
				topCalories = currentCalories
			}
			currentCalories = 0
		} else {
			calories,_ := strconv.Atoi(line)
			currentCalories += calories	
		}
	
	} 
	
	if topCalories < currentCalories {
		topCalories = currentCalories
	}
	return topCalories
}

func step2(data []string) int {
	topCalories := []int{0,0,0}
	var currentCalories int = 0
	for _,line := range data {
		if (line == "") {
			topCalories = updateTopCalories(topCalories,currentCalories)
			currentCalories = 0
		} else {
			calories,_ := strconv.Atoi(line)
			currentCalories += calories	
		}
			sort.Ints(topCalories)
	} 
	
	topCalories = updateTopCalories(topCalories,currentCalories)

	var total int = 0
	for _,calories := range topCalories{
		total += calories
	}
	return total
}

func updateTopCalories(topCalories []int, currentCalories int) []int {
	if currentCalories > topCalories[0] {
		topCalories = append(topCalories, currentCalories)
		topCalories = topCalories[1:]
	}

	return topCalories
}

File: 2022/tom/day-1/rust/src/main.rs
use std::{
    fs::File,
    io::{prelude::*, BufReader},
    path::Path,
};
use std::env;

fn lines_from_file(filename: impl AsRef<Path>) -> Vec<String> {
    let file = File::open(filename).expect("no such file");
    let buf = BufReader::new(file);
    buf.lines()
        .map(|l| l.expect("Could not parse line"))
        .collect()
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let data;
    let test_flag: String = "test".to_string();
    if args.contains(&test_flag) {
        data = lines_from_file("../test.txt");
    } else {
        data = lines_from_file("../input.txt");
    }

    let step1 = step1(&data);

    println!("Step 1: {}", step1);
    let step2= step2(&data);

    println!("Step 2: {}", step2);
}

fn step1(data: &Vec<String>) -> i64 {
	let mut top_calories = 0;
	let mut current_calories = 0;
	for line in data {
		if line == "" {
			if top_calories < current_calories {
				top_calories = current_calories;
			}
			current_calories = 0;
		} else {
			let calories: i64 = line.parse().unwrap();
			current_calories += calories	;
		}
	
	} 
	
	if top_calories < current_calories {
		top_calories = current_calories;
	}
	return top_calories;
}

fn step2(data: &Vec<String>) -> i64 {
	let mut top_calories = vec![0,0,0];
	let mut current_calories = 0;
	for line in data {
		if line == "" {
			top_calories = update_top_calories(top_calories,current_calories);
			current_calories = 0;
		} else {
			let calories: i64 = line.parse().unwrap();
			current_calories += calories	;
		}
	}
	
	top_calories = update_top_calories(top_calories,current_calories);

	let mut total = 0;
    for calories in top_calories {
		total += calories;
	}
	return total;
}

fn update_top_calories(mut top_calories: Vec<i64>,current_calories: i64) -> Vec<i64>{
    top_calories.sort_by(|a, b| b.cmp(a));
    if current_calories > top_calories[2] {
        top_calories.pop();
        top_calories.push(current_calories)
    }
    return top_calories;
  }